### PR TITLE
Refactor ligand-based models into a library-like project

### PR DESCRIPTION
**Pull request by jaimergp**
## Description
This PR aims to refactor @t-kimber's scripts for ligand-based models into a library-consumable infrastructure.

## Todos

Library:

- [ ] Bring in featurizers: 
    - [ ] Normalize Hash Featurizer for protein 
    - [ ] Dominique's kinase fingerprint
- [ ] Bring in models
- [ ] Design `Ligand` and `Protein` helper objects for easy handling of raw data and their potential featurizations. `Ligand` object can subclass `openforcefield.topology.Molecule`, but I am not sure how we should deal with `Protein`...
- [ ] Provide `DatasetConsumers` (or `*Builder`, `*Handler`, etc) base class and corresponding subclasses for KINOMEscan, ChEMBL, etc.
- [ ] Add unit tests
- [ ] Make sure docstrings are there and are accurate @t-kimber 

Scripts/workflows:

- [ ] Add CLI-enabled scripts mimicking existing scripts
- [ ] Make sure results are consistent between original and refactored codebases.

## Status
- [ ] Ready to go

### Ligand
|                        | Class |  Docstrings | Tests |
| ------------- |:-----:| :-----------:| :------: |
|     Morgan      | :heavy_check_mark: | :heavy_check_mark: | :heavy_check_mark: |
|    SMILES       | :heavy_check_mark: | :heavy_check_mark: | :heavy_check_mark: |
|       Graph      | :heavy_check_mark: | :heavy_check_mark: | :heavy_check_mark: |

### Protein
|                        | Class |  Docstrings | Tests |
| ------------- |:-----:| :-----------:| :------: |
| Hash              | :heavy_check_mark:  |:heavy_check_mark:  | :heavy_check_mark: |
| Composition  | :heavy_check_mark: | :heavy_check_mark: | :heavy_check_mark: |
| Sequence     | :heavy_check_mark: | :heavy_check_mark: | :heavy_check_mark: |

### Models
|                        | Class |  Docstrings | Tests |
| ------------- |:-----:| :-----------:| :------: |
| XGB        |     :x:    |       :x:        |   :x:    |
| DNN               | :heavy_check_mark: | :heavy_check_mark: | :x: |
| CNN               | :heavy_check_mark: | :heavy_check_mark: | :x: |
|MPNN             |     :x:     |        :x:       |       :x:    |

### CODE REVIEW

File: kinoml/workflows/chembl22_cnn.py [hunk 1]
```diff
@@ -0,0 +1,188 @@
+'''
+This script uses the pre-processed data file df_reduced.pkl that is generated from the ChEMBL22 data set.
+
+Type of target used: single-target (i.e. ligand-based model)
+Type of featurizer for the ligand: One Hot Encoding of SMILES
+Type of model: a Convolutional Neural Network (CNN) using Pytorch
+
+Please run in terminal the following command:
+$ export PYTHONPATH=$(dirname $(pwd))
```
Comment (jaimergp) [inline kinoml/workflows/chembl22_cnn.py]: We don't want to mess up with `PYTHONPATH` (or `sys.path` runtime modifications). Development should happen with the package installed in development mode:

```
pip install -e .
```

File: kinoml/workflows/chembl22_cnn.py [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+
+if __name__ == '__main__':
+
+    # Load pre-processed data
+    p = Path(__file__).resolve()
+    data = pd.read_pickle(f'{p.parents[1]}/data/df_reduced.pkl')
+
+    # Set threshold for activity
+    c = 6.3
+    print(f'Threshold at {c}')
```
Comment (jaimergp) [inline kinoml/workflows/chembl22_cnn.py]: If we are aiming at providing library-like components, `print()` shouldn't be used because downstream users won't be able to change the verbosity easily (they would need to patch `sys.stdout` or weird hacks like that), so it's better if we use  `logging`.

File: kinoml/workflows/chembl22_cnn.py [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+
+from ml.models_pytorch import CNN
+from features.ligand import OneHotSMILESFeaturizer
+
+
+if __name__ == '__main__':
+
+    # Load pre-processed data
+    p = Path(__file__).resolve()
+    data = pd.read_pickle(f'{p.parents[1]}/data/df_reduced.pkl')
```
Comment (jaimergp) [inline kinoml/workflows/chembl22_cnn.py]: We need to figure out a way to use entry points for data files. At least a helper function, applicable all across the library, should be used here (something like `get_data_file('df_reduced.pkl')`).

File: kinoml/workflows/chembl22_cnn.py [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+            print('---Predict')
+            # Test trained model on the test set
+            outputs_test = torch.squeeze(model(x_test))
+
+            # Transform target vectors into numpy array for scores
+            y_pred = outputs_test.detach().numpy()
+            y_test = y_test.detach().numpy()
+
+            # Transform probability of belonging to one class into 0 or 1 value at threshold 0.5
+            y_pred =(y_pred>0.5)
```
Comment (jaimergp) [inline kinoml/workflows/chembl22_cnn.py]: Parenthesis not needed here.

File: kinoml/workflows/chembl22_cnn.py [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+        acc_mean = np.mean(acc_per_fold)
+        conf_matrix_mean = np.mean(np.array(conf_matrix_per_fold), axis=0)
+
+        # Std on all K folds
+        acc_std = np.std(acc_per_fold)
+        conf_matrix_std = np.std(np.array(conf_matrix_per_fold), axis=0)
+
+        print(f' Metrics of CV for kinase {kinase} ')
+        print(f' Accuracy : Mean : {round(acc_mean, 2)} and std : {round(acc_std, 2)} ')
+        print(f' TN : Mean : {round(conf_matrix_mean[0], 2)} and std : {round(conf_matrix_std[0], 2)} ')
```
Comment (jaimergp) [inline kinoml/workflows/chembl22_cnn.py]: You can use format notation for floats to easily get the rounded quantity. For example, `f'{acc_mean:.2f}` means "print `acc_mean` as a float with two decimal positions. Check the _new_ style in [this website](https://pyformat.info/).

File: kinoml/workflows/chembl22_cnn.py [hunk 6]
```diff
@@ -0,0 +1,188 @@
...
+        conf_matrix_std_per_kinase.append(conf_matrix_std)
+
+
+    # Save results to a pickle table in folder cv_results
+    results_cv = pd.DataFrame(data=[acc_mean_per_kinase, acc_std_per_kinase,
+                                    conf_matrix_mean_per_kinase, conf_matrix_std_per_kinase,
+                                    zero_count_per_kinase])
+    results_cv.columns = [data.columns[:(-2-k)]]
+    results_cv.index = ['Acc Mean', 'Acc Std', 'Conf Mean', 'Conf Std', 'Zero Count']
+    results_cv = results_cv.to_pickle('CNN_class_results_cv.pkl')
```
Comment (jaimergp) [inline kinoml/workflows/chembl22_cnn.py]: Outputs should be parameterized too.

Comment (jaimergp) [reply kinoml/workflows/chembl22_cnn.py]: (You can always set a default keyword argument in the hypothetical `main` function).

File: kinoml/workflows/chembl22_cnn.py (hunk 3 above)
Comment (jaimergp) [reply kinoml/workflows/chembl22_cnn.py]: By the way, we should try to parameterize the input file in case somebody wants to try a different dataset (with similar data in it), without having to patch this one.

File: kinoml/features/ligand.py [hunk 7]
```diff
@@ -0,0 +1,179 @@
...
+        ohe_matrix = np.zeros((len(self.DICTIONARY), len(self.molecule)))
+        for i, dict_char in enumerate(self.DICTIONARY):
+            for j, molecule_char in enumerate(self.molecule):
+                if dict_char == molecule_char:
+                    ohe_matrix[i,j] = 1
+        if self.pad_up_to is not None:
+            return np.pad(ohe_matrix, ((0,0), (0, self.pad_up_to-len(self.molecule))), mode='constant')
+        else: 
+            return ohe_matrix
+#        smiles = self.molecule.to_smiles().replace("Cl", "L").replace("Br", "R").replace("@@", "$")
```
Comment (jaimergp) [inline kinoml/features/ligand.py]: You can just remove these commented out lines.

File: kinoml/ml/models_pytorch.py [hunk 8]
```diff
@@ -0,0 +1,56 @@
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+
+class DNN(nn.Module):
```
Comment (jaimergp) [inline kinoml/ml/models_pytorch.py]: We'll need some docstrings in this classes!

File: kinoml/workflows/chembl22_dnn.py [hunk 9]
```diff
@@ -0,0 +1,189 @@
+'''
```
Comment (jaimergp) [inline kinoml/workflows/chembl22_dnn.py]: Comments left in the CNN workflows also apply here.

File: kinoml/workflows/ohe_seq_feat.py [hunk 10]
```diff
@@ -0,0 +1,35 @@
+"""
```
Comment (jaimergp) [inline kinoml/workflows/ohe_seq_feat.py]: Comments left in the CNN workflow also apply here.

File: kinoml/features/protein.py [hunk 11]
```diff
@@ -0,0 +1,110 @@
...
+        =======
+        np.matrix
+            One hot encoding of the sequence, with shape (``len(ALL_AMINOACIDS)``, ``len(self.molecule.sequence)``).
+    """
+
+    def __init__(self, amino_acid_seq):
+        self.amino_acid_seq = amino_acid_seq
+        self.len_seq = len(amino_acid_seq)
+    
+    def from_seq2oh(self):
```
Comment (jaimergp) [inline kinoml/features/protein.py]: Is this considered a featurizer too? In that case we should stick to the API we are trying to build (i.e. subclass `_BaseFeaturizer` and redefine their methods).

File: kinoml/features/ligand.py [hunk 12]
```diff
@@ -0,0 +1,179 @@
...
+        Featurizes ``self.molecule`` as a Morgan Fingerprint using RDKit
+
+        Returns
+        ========
+        np.array
+            Morgan fingerprint of radius 2 of molecule, with shape ``nbits``. 
+        """
+        from rdkit import Chem
+        from rdkit.Chem.AllChem import GetMorganFingerprintAsBitVect
+        #m = self.molecule.to_rdkit()
```
Comment (jaimergp) [inline kinoml/features/ligand.py]: We need to try to use those `Molecule` objects that are not yet defined. The `to_rdkit` logic should be encapsulated in those objects so the featurizer doesn't have to assume much (right now it's assuming that the ligand will always be a SMILES string). The `Ligand` objects are empty for now, but we can start filling them in as we find the need (and this is a good one!).

**Comment by t-kimber** [comment]
### Ligand

|                        | Class |  Docstrings | Tests |
| ------------- |:-----:| -----------:| ------ |
|     Morgan     | :heavy_check_mark:    |         :heavy_check_mark:          |     :heavy_check_mark:  |
| SMILES OHE |       :heavy_check_mark:   |     :heavy_check_mark:            |      :heavy_check_mark:    |
|       Graph      |       :x:   |    :x:              |       :x:    |


### Protein
|                        | Class |  Docstrings | Tests |
| ------------- |:-----:| -----------:| ------ |
|    Hash            |     :x:    |       :x:        |   :x:    |
| Comp.             | :x:        |    :x:             |     :x:    |
| Seq. OHE        |      :x:   |    :x:             |      :x:    |
|Structural FP   |     :x:     |        :x:       |       :x:    |

**Comment by t-kimber** [comment]
**TODO:** change naming in MorganFingerprintFeaturizer class to avoid having molecule.molecule.

**Comment by jaimergp** [comment]
Superseded by #8
